marathon/src/main/scala/io/buoyant/marathon/v2/Api.scala

```diff
@@ -101,21 +133,119 @@ object Api {
   }
 }
 
-private class AppIdApi(client: Api.Client, apiPrefix: String)
+private class AppIdApi(client: Api.Client, apiPrefix: String, auth: Option[Api.AuthRequest])
   extends Api
   with Closable {
 
   import Api._
 
   def close(deadline: Time) = client.close(deadline)
 
-  def getAppIds(): Future[Api.AppIds] = {
-    val req = http.Request(s"$apiPrefix/apps")
-    Trace.letClear(client(req)).flatMap(rspToApps(_))
-  }
+  private[this] val state = Var[Activity.State[Option[String]]](Activity.Ok(None))
+  private[this] val authToken = Activity(state)
+
+  private[this] def refreshToken(): Future[String] =
+    auth match {
+      case None => Future.value("")
+      case Some(auth) => synchronized {
+        state.sample() match {
+          case Activity.Pending =>
+            authToken.values.toFuture.flatMap {
+              case Return(Some(token)) => Future.value(token)
+              case Return(None) => Future.value("")
+              case Throw(e) =>
+                log.error(s"authToken threw ${e}")
+                Future.exception(e)
+            }
+          case _ =>
```

Comment:
could potentially use a comment here about how we want to force refresh the token, even if one already exists (it might be expired).